### PR TITLE
Realease more users

### PR DESCRIPTION
**Pull request by Casassarnau**
## What this PR does / Why we need it
A few sentences describing the overall goals of the pull request's commits.

## Which issue(s) this PR fixes (optional) 

References #288 

## Special notes for your reviewer (optional)
User type added and adapted to Hackassistant. New ways to register done as well.
signup/hacker -> Registers hackers
signup/volunteer -> Registers volunteer
signup/mentor -> Registers mentor
signup/sponsor -> Registers sponsor with SPONSOR_TOKEN
signup/else -> Type not found

## Some questions
- [x] I have read the contributing guidelines
- [x] I abide by this repository Code of Conduct
- [x] I understand that my PR won't be merged until GitHub gives a "green light"

## Additional Notes (optional)
Do you want to add anything else? We :heart: to hear your opinions!

### CODE REVIEW

File: app/hackathon_variables.py [hunk 1]
```diff
@@ -82,3 +82,5 @@
 
 # Enable dubious separate pipeline (disabled by default)
 DUBIOUS_ENABLED = True
+
+SUPPORTED_RESUME_EXTENSIONS = ['.pdf']
```
Comment (casassg) [inline app/hackathon_variables.py]: Why is this here?

File: user/migrations/0009_auto_20200321_0441.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        ),
+        migrations.RemoveField(
+            model_name='user',
+            name='is_volunteer',
+        ),
+        migrations.AddField(
+            model_name='user',
+            name='type',
+            field=models.CharField(choices=[('H', 'Hacker'), ('M', 'Mentor'), ('S', 'Sponsor'), ('V', 'Volunteer'), ('O', 'Organizer')], default='H', max_length=2),
+        ),
```
Comment (casassg) [inline user/migrations/0009_auto_20200321_0441.py]: Could you change the name of the migration to something more named (instead of auto....)

Also, if you can, would it be possible to do use https://docs.djangoproject.com/en/3.0/ref/migration-operations/#runsql to update the DB from the previous to new state? (if `is_organizer` set type to `organizer`.....) Or https://docs.djangoproject.com/en/3.0/ref/migration-operations/#runpython

File: user/models.py [hunk 3]
```diff
@@ -4,15 +4,38 @@
...
+    (USR_VOLUNTEER, 'Volunteer'),
+    (USR_ORGANIZER, 'Organizer'),
+]
+
+USR_URL_SPONSOR = 'sponsor'
+USR_URL_TYPE = {
+    'hacker': USR_HACKER,
+    'volunteer': USR_VOLUNTEER,
+    'mentor': USR_MENTOR,
+    USR_URL_SPONSOR: USR_SPONSOR
```
Comment (casassg) [inline user/models.py]: Isnt this label.lower()?

File: user/models.py [hunk 4]
```diff
@@ -110,3 +134,11 @@ def is_staff(self):
...
     def has_dubious_acces(self):
         return self.can_review_dubious or self.is_director
+
+    @property
+    def is_organizer(self):
+        return self.type == USR_ORGANIZER
+
+    @property
+    def is_volunteer(self):
+        return self.type == USR_VOLUNTEER
```
Comment (casassg) [inline user/models.py:149]: 👍

File: user/urls.py [hunk 5]
```diff
@@ -5,7 +5,7 @@
 urlpatterns = [
     url(r'^login/$', views.login, name='account_login'),
     url(r'^callback/(?P<provider>[0-9A-Za-z_\-]+)/$', views.callback, name='callback'),
-    url(r'^signup/$', views.signup, name='account_signup'),
+    url(r'^signup/(?P<u_type>[a-z_\-]{1,10})/$', views.signup, name='account_signup'),
```
Comment (casassg) [inline user/urls.py]: Add a redirect for `signup/` to `signup/hacker` for backwards compatibility pls.

File: user/views.py [hunk 6]
```diff
@@ -61,12 +64,15 @@ def signup(request):
...
+                user = models.User.objects.create_user(email=email, password=password, name=name, u_type=u_type)
                 user = auth.authenticate(email=email, password=password)
                 auth.login(request, user)
                 return HttpResponseRedirect(reverse('root'))
     else:
-        form = forms.RegisterForm()
+        if u_type == models.USR_URL_SPONSOR:
+            form = forms.RegisterSponsorForm()
+        else:
+            form = forms.RegisterForm()
```
Comment (casassg) [inline user/views.py]: Probably store a dict for given a u_type what Form instance they should use.

**Review by casassg — CHANGES_REQUESTED** [review]

File: app/hackathon_variables.py (hunk 1 above)
Comment (Casassarnau) [reply app/hackathon_variables.py]: don't really know :/

File: user/migrations/0009_auto_20200321_0441.py (hunk 2 above)
Comment (Casassarnau) [reply user/migrations/0009_auto_20200321_0441.py]: Aren't migrations named automatically?

File: user/models.py (hunk 3 above)
Comment (Casassarnau) [reply user/models.py]: This is in order to change from 'hacker' to 'H'. I could take the first char and upper() but then if in the future we have more users with 2 chars, this won't work. And I thougth it is more readable like this

File: user/migrations/0009_auto_20200321_0441.py (hunk 2 above)
Comment (casassg) [reply user/migrations/0009_auto_20200321_0441.py]: Yes, but you can name them by using the `-n` option.

Also, mind to provide the `reverse_sql` (SQL to reverse the migration). Mostly for completeness (if someone migrates and has issues can rollback safely.

File: user/migrations/0009_user_types_20200321_0441.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    operations = [
+        migrations.AddField(
+            model_name='user',
+            name='type',
+            field=models.CharField(
+                choices=[('H', 'Hacker'), ('M', 'Mentor'), ('S', 'Sponsor'), ('V', 'Volunteer'), ('O', 'Organizer')],
+                default='H', max_length=2),
+        ),
+        migrations.RunSQL(["UPDATE user_user SET type='V' where is_volunteer;"]),
+        migrations.RunSQL(["UPDATE user_user SET type='O' where is_organizer;"]),
```
Comment (casassg) [inline user/migrations/0009_user_types_20200321_0441.py]:
```suggestion
        migrations.RunSQL(["UPDATE user_user SET type='V' where is_volunteer;"], reverse_sql="UPDATE user_user SET is_volunteer=1 where type='V';"),
        migrations.RunSQL(["UPDATE user_user SET type='O' where is_organizer;"], reverse_sql="UPDATE user_user SET is_organizer=1 where type='O';"),
```

File: user/models.py (hunk 3 above)
Comment (casassg) [reply user/models.py]: USR_TYPE has `Hacker` and USR_URL_TYPE has `hacker`. I wonder if that can be collapsed.

File: user/views.py (hunk 6 above)
Comment (Casassarnau) [reply user/views.py]: I would have to create uno for the () and another for the post. I think an if it's better

File: user/migrations/0009_user_types_20200321_0441.py (hunk 7 above)
Comment (Casassarnau) [reply user/migrations/0009_user_types_20200321_0441.py]: Already done

File: user/urls.py (hunk 5 above)
Comment (Casassarnau) [reply user/urls.py]: Done

File: user/models.py (hunk 3 above)
Comment (Casassarnau) [reply user/models.py]: Done

File: app/hackathon_variables.py (hunk 1 above)
Comment (Casassarnau) [reply app/hackathon_variables.py]: I will have it in mind on the last merge.

File: user/forms.py [hunk 8]
```diff
@@ -47,6 +53,26 @@ def clean_password2(self):
...
+        return self.cleaned_data
+
+
+class RegisterSponsorForm(RegisterForm):
+    token = forms.CharField(label='Registration code', max_length=50)
+    field_order = ['name', 'email', 'password', 'password2', 'token']
+
+    def clean_token(self):
+        token = self.cleaned_data.get("token")
+        if token != settings.SPONSOR_TOKEN:
```
Comment (casassg) [inline user/forms.py]: This will fail if SPONSOR_TOKEN is not defined.

File: user/views.py [hunk 9]
```diff
@@ -47,12 +47,15 @@ def login(request):
...
+def signup(request, u_type):
     if request.user.is_authenticated:
         return HttpResponseRedirect(reverse('root'))
     # if this is a POST request we need to process the form data
     if request.method == 'POST':
-        form = forms.RegisterForm(request.POST)
+        if u_type == models.USR_URL_SPONSOR:
+            form = forms.RegisterSponsorForm(request.POST, type=u_type)
+        else:
+            form = forms.RegisterForm(request.POST, type=u_type)
```
Comment (casassg) [inline user/views.py]:
```suggestion
       form_by_user = {
           models.USR_URL_SPONSOR: forms.RegisterSponsorForm,
        }
        form = form_by_user.get(u_type, forms.RegisterForm)(request.POST, type=u_type)
```

**Review by casassg — APPROVED** [review]
